File: backend/app/connectors/executors/nexplane_agent/containerize_retire.py
"""Executor for agent_containerize_retire change type."""


async def execute(parameters: dict, asset_ids: list, connector) -> dict:
    """Dispatch containerize_retire to the registered Nexplane agent."""
    from app.connectors.executors.nexplane_agent._dispatch import dispatch_agent_job

    systemd_unit = parameters.get("systemd_unit", "")
    if not systemd_unit:
        raise ValueError("Missing required parameter: systemd_unit")

    return await dispatch_agent_job(
        command="containerize_retire",
        parameters={
            "systemd_unit": systemd_unit,
            "dry_run": bool(parameters.get("dry_run", False)),
        },
        asset_ids=list(asset_ids),
        timeout_seconds=120,
    )


async def rollback(parameters: dict, execution_result: dict, connector) -> dict:
    """Rollback: restart the legacy service."""
    from app.connectors.executors.nexplane_agent._dispatch import dispatch_agent_job

    systemd_unit = parameters.get("systemd_unit") or execution_result.get("systemd_unit", "")
    if not systemd_unit:
        return {"rolled_back": False, "note": "No systemd_unit for rollback"}

    result = await dispatch_agent_job(
        command="containerize_retire",
        parameters={"systemd_unit": systemd_unit, "dry_run": False},
        asset_ids=list(asset_ids) if (asset_ids := parameters.get("asset_ids", [])) else [],
        timeout_seconds=60,
        rollback=True,
    )
    return result

File: agent/commands/containerizeretire/containerizeretire.go
package containerizeretire

import (
	"fmt"
	"os/exec"
	"strings"
)

// ContainerizeRetireExecute stops and disables the legacy systemd service.
// Parameters:
//   - systemd_unit (string, required): e.g. "myapp.service"
//   - dry_run (bool, optional): if true, skip actual systemctl calls
func ContainerizeRetireExecute(params map[string]any) (map[string]any, error) {
	unit, ok := params["systemd_unit"].(string)
	if !ok || unit == "" {
		return nil, fmt.Errorf("missing required parameter: systemd_unit")
	}
	unit = strings.TrimSuffix(unit, ".service") + ".service"

	dryRun, _ := params["dry_run"].(bool)

	result := map[string]any{
		"action":           "containerize_retire",
		"systemd_unit":     unit,
		"dry_run":          dryRun,
		"service_stopped":  false,
		"service_disabled": false,
	}

	if dryRun {
		result["note"] = "dry_run: no changes made"
		return result, nil
	}

	stopOut, err := exec.Command("systemctl", "stop", unit).CombinedOutput()
	if err != nil {
		return result, fmt.Errorf("systemctl stop %s failed: %w — output: %s", unit, err, stopOut)
	}
	result["service_stopped"] = true

	disableOut, err := exec.Command("systemctl", "disable", unit).CombinedOutput()
	if err != nil {
		result["disable_warning"] = fmt.Sprintf("systemctl disable failed: %s", disableOut)
	} else {
		result["service_disabled"] = true
	}

	return result, nil
}

// ContainerizeRetireRollback restarts the legacy service.
func ContainerizeRetireRollback(params map[string]any) (map[string]any, error) {
	unit, ok := params["systemd_unit"].(string)
	if !ok || unit == "" {
		return map[string]any{"rolled_back": false, "note": "no systemd_unit in rollback params"}, nil
	}
	unit = strings.TrimSuffix(unit, ".service") + ".service"

	out, err := exec.Command("systemctl", "start", unit).CombinedOutput()
	if err != nil {
		return map[string]any{
			"rolled_back": false,
			"error":       fmt.Sprintf("systemctl start %s failed: %s", unit, out),
		}, nil
	}
	return map[string]any{
		"rolled_back":       true,
		"systemd_unit":      unit,
		"service_restarted": true,
	}, nil
}
